rock-n-rollup/src/core/runtime.rs

```diff
@@ -104,132 +49,127 @@ pub trait Runtime: 'static {
         match res {
-            0 => Ok(()),
-            _ => Err(()),
+            Ok(_) => Ok(()),
+            Err(_) => Err(print!("Error store_write")),
         }
     }
 
-    fn store_read(&mut self, path: &str, offset: usize, size: usize) -> Option<Vec<u8>> {
-        if !self.store_is_present(path) {
-            return None;
-        }
+    fn store_move(&mut self, from: &str, to: &str) -> Result<(), ()> {
+        let from = OwnedPath::try_from(from.to_string()).map_err(|_| ())?;
+        let to = OwnedPath::try_from(to.to_string()).map_err(|_| ())?;
 
-        let ptr = path.as_ptr();
-        let path_len = path.len();
-        let mut buffer = Vec::with_capacity(size);
-        let dst = buffer.as_mut_ptr();
-        unsafe {
-            let _ = store_read(ptr, path_len, offset, dst, size);
-            buffer.set_len(size);
+        let res = self.host.store_move(&from, &to);
+        match res {
+            Ok(_) => Ok(()),
+            Err(_) => Err(print!("Error store_move")),
         }
-
-        Some(buffer)
     }
 
-    fn store_write(&mut self, path: &str, data: &[u8], at_offset: usize) -> Result<(), ()> {
-        let res = unsafe {
-            let path_len = path.len();
-            let path = path.as_ptr();
-            let num_bytes = data.len();
-            let src = data.as_ptr();
-            store_write(path, path_len, at_offset, src, num_bytes)
-        };
+    fn reveal_preimage(&mut self, hash: &[u8; PREIMAGE_HASH_SIZE]) -> Result<Vec<u8>, ()> {
+        let payload: Vec<u8> = Vec::with_capacity(MAX_MESSAGE_SIZE);
+        let mut payload: Vec<u8> = payload.to_vec();
+        let res: Result<usize, _> = self.host.reveal_preimage(hash, &mut payload);
         match res {
-            0 => Ok(()),
-            err => {
-                self.write_debug(&format!("error store_write_raw: {}\n", err));
-                Err(())
-            }
+            Ok(bytes) => Ok(bytes.to_le_bytes().to_vec()),
+            Err(_) => Err(print!("Error reveal_preimage")),
         }
     }
-    fn reveal_preimage(&mut self, hash: &[u8; PREIMAGE_HASH_SIZE]) -> Result<Vec<u8>, ()> {
-        let max_size = 4096;
-        let mut payload = Vec::with_capacity(MAX_MESSAGE_SIZE as usize);
-
-        let u8_size = u8::try_from(PREIMAGE_HASH_SIZE).unwrap();
-
-        unsafe {
-            let size = reveal_preimage(hash.as_ptr(), u8_size, payload.as_mut_ptr(), max_size);
-            if size < 0 {
-                Err(())
-            } else {
-                let size = usize::try_from(size).unwrap();
-                payload.set_len(size);
-                Ok(payload)
+
+    fn store_is_present(&mut self, path: &str) -> bool {
+        let path = OwnedPath::try_from(path.to_string()).map_err(|_| ());
+        match path {
+            Ok(path) => {
+                let res = unsafe { self.host.store_has(&path) };
+                match res {
+                    Ok(Some(value_type)) => match value_type {
```

Comment:
You don't need to match on the value_type, if it's a Some you can directly returns true

Reply:
I optimize the code at this [commit](f62e693698d8b622c3351f686d7de6cd33b20721)